components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt

```diff
@@ -0,0 +1,332 @@
+/*
+  Copyright (C) 2013-2022 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.client
+
+import com.hotels.styx.api.HttpHeaderNames
+import com.hotels.styx.api.extension.loadbalancing.spi.LoadBalancer
+import com.hotels.styx.api.extension.retrypolicy.spi.RetryPolicy
+import com.hotels.styx.api.extension.service.StickySessionConfig
+import com.hotels.styx.metrics.CentralisedMetrics
+import com.hotels.styx.api.LiveHttpRequest
+import com.hotels.styx.api.HttpInterceptor
+import com.hotels.styx.api.HttpMethod
+import com.hotels.styx.api.HttpResponseStatus
+import com.hotels.styx.api.Id
+import com.hotels.styx.api.LiveHttpResponse
+import reactor.core.publisher.Flux
+import com.hotels.styx.api.exceptions.NoAvailableHostsException
+import com.hotels.styx.api.ResponseEventListener
+import com.hotels.styx.api.extension.Origin
+import com.hotels.styx.api.extension.RemoteHost
+import com.hotels.styx.api.extension.service.RewriteRule
+import com.hotels.styx.client.stickysession.StickySessionLoadBalancingStrategy
+import com.hotels.styx.client.stickysession.StickySessionCookie
+import com.hotels.styx.client.retry.RetryNTimes
+import org.reactivestreams.Publisher
+import org.slf4j.LoggerFactory
+import java.lang.StringBuilder
+import java.util.Objects.nonNull
+import java.util.Optional
+
+/**
+ * A configurable HTTP client that uses connection pooling, load balancing, etc.
+ */
+class StyxBackendServiceClient(
+    private val id: Id,
+    rewriteRules: List<RewriteRule>,
+    private val originStatsFactory: OriginStatsFactory,
+    private val originsRestrictionCookieName: String?,
+    private val stickySessionConfig: StickySessionConfig,
+    private val originIdHeader: CharSequence,
+    private val loadBalancer: LoadBalancer,
+    private val retryPolicy: RetryPolicy,
+    private val metrics: CentralisedMetrics,
+    private val overrideHostHeader: Boolean
+) : BackendServiceClient {
+    private val rewriteRuleset: RewriteRuleset = RewriteRuleset(rewriteRules)
+
+    private constructor(builder: Builder): this(
+        id = builder.id,
+        originStatsFactory = builder.originStatsFactory!!,
+        loadBalancer = builder.loadBalancer!!,
```

Comment:
Do you think we should add a check in .build() for loadBalancer not being null, as you did for metrics? Not a huge fan of using !! in production code, in any case...
You could potentially do like
```
originStatsFactory = builder?.originStatsFactory ?: somethingElseHere
```
even though you know that should never happen.

Reply:
@ryanellison yeah I could definitely do that; @kvosper I'll tag you to get your thoughts too since you suggested the !!

I already check metrics and originStatsFactory being null in .build(), do you think I should also add a check for loadBalancer there **and** do something like your suggestion for **all three** of metrics, originStatsFactory, and loadBalancer here in the constructor to remove the need of !!?

Reply:
I guess if @kvosper is alright with !! I'd just add an additional check into .build() to protect against this being null, since we're saying it's definitely not with !!.
The tricky think about providing a default for originStatsFactory is the the factory for one of those requires non-null "metrics" which starts to get a little ugly putting it all here...
For reference, in your builder:
```
OriginStatsFactory.CachingOriginStatsFactory(metrics)
```

Reply:
Cool, I'll add a null check in .build() for loadBalancer and leave the !! as is for now then

Reply:
^ pushed changes, squashed into the previous commit

Reply:
After thinking about it, `!!` is not the best choice, as this simply throws a `NullPointerException` if something is null. It would be better to use Kotlin's `requireNotNull` like so `requireNotNull(builder.loadBalancer) { "loadBalancer is required" } `.

I may have confused this with java's `Objects.requireNonNull` earlier (which should not be used in Kotlin), but using Kotlin's functions is OK.